### PR TITLE
Create default empty branch head commits

### PR DESCRIPTION
**Pull request by Tryneus**
This simplifies a lot of handling in PFS if we can assume that there is always a branch head.  In CreateBranch and SquashCommit, we ensure that a branch head always exists.

### CODE REVIEW

File: src/server/pfs/cmds/cmds.go [hunk 1]
```diff
@@ -669,21 +670,35 @@ Any pachctl command that can take a Commit ID, can take a branch name instead.`,
...
+						return err
+					}
+				}
+			}
+
 			return txncmds.WithActiveTransaction(c, func(c *client.APIClient) error {
 				_, err := c.PfsAPIClient.CreateBranch(
 					c.Ctx(),
 					&pfs.CreateBranchRequest{
-						Head:       branch.Repo.NewCommit("", head),
```
Comment (Tryneus) [inline src/server/pfs/cmds/cmds.go:686]: This was broken - it was specifying a head commit even if we weren't updating the branch head.

File: src/server/pfs/server/driver.go [hunk 2]
```diff
@@ -671,23 +684,6 @@ func (d *driver) aliasCommit(txnCtx *txncontext.TransactionContext, parent *pfs.
 	return commitInfo, nil
 }
 
-// writeFinishedCommit writes these changes to etcd:
-// 1) it closes the input commit (i.e., it writes any changes made to it and
-//    removes it from the open commits)
-// 2) if the commit is the new HEAD of master, it updates the repo size
-func (d *driver) writeFinishedCommit(sqlTx *sqlx.Tx, commitInfo *pfs.CommitInfo) error {
```
Comment (Tryneus) [inline src/server/pfs/server/driver.go:678]: this function was obsolete (we no longer track the repo size in the repo info)

File: src/server/pfs/server/driver.go [hunk 3]
```diff
@@ -580,8 +580,21 @@ func (d *driver) finishCommit(txnCtx *txncontext.TransactionContext, commit *pfs
 	if description != "" {
 		commitInfo.Description = description
 	}
+	// Check that the parent commit is not still open (which can happen in output branches)
+	if commitInfo.ParentCommit != nil {
```
Comment (Tryneus) [inline src/server/pfs/server/driver.go:584]: Unrelated fix for finishing a commit on an unfinished parent which should never happen.

File: src/server/pfs/pfs.go [hunk 4]
```diff
@@ -289,3 +283,12 @@ func IsCommitOnOutputBranchErr(err error) bool {
 	}
 	return commitOnOutputBranchRe.MatchString(err.Error())
 }
+
+// IsParentNotFinishedErr returns true if the err is due to an attempt to
```
Comment (brycemcanally) [inline src/server/pfs/pfs.go:287]: I don't think this comment is correct.

File: src/server/pfs/server/driver.go [hunk 5]
```diff
@@ -791,17 +787,14 @@ func (d *driver) propagateBranches(txnCtx *txncontext.TransactionContext, branch
 			if err != nil {
 				return err
 			}
-			if provOfSubvBI.Head == nil || provOfSubvBI.Branch.Repo.Type == pfs.SpecRepoType {
```
Comment (brycemcanally) [inline src/server/pfs/server/driver.go:794]: What was the reason for filtering out the spec repo originally?

File: src/server/pfs/server/driver.go [hunk 6]
```diff
@@ -791,17 +787,14 @@ func (d *driver) propagateBranches(txnCtx *txncontext.TransactionContext, branch
 			if err != nil {
```
Comment (brycemcanally) [inline src/server/pfs/server/driver.go:787]: This comment is intended for the `propagateBranches` function comment (I guess you cannot comment on collapsed lines). We should probably remove the reference to a nil HEAD.

File: src/server/pfs/server/driver.go [hunk 7]
```diff
@@ -791,17 +787,14 @@ func (d *driver) propagateBranches(txnCtx *txncontext.TransactionContext, branch
...
 				needsCommit = true
 				break
 			}
 		}
 
 		// If there are no upstream commits for this Commitset and no commit in this branch, we can skip
-		if !needsCommit && (subvBI.Head == nil || subvBI.Head.ID != txnCtx.CommitsetID) {
+		if !needsCommit && subvBI.Head.ID != txnCtx.CommitsetID {
 			continue
 		}
```
Comment (brycemcanally) [inline src/server/pfs/server/driver.go:800]: There is a nil HEAD check in the provenance iteration below.

File: src/server/pfs/server/driver.go [hunk 8]
```diff
@@ -1915,26 +1911,30 @@ func isNotFoundErr(err error) bool {
...
 	if pfsdb.BranchKey(provBranch) == pfsdb.BranchKey(branchInfo.Branch) {
 		return errors.Errorf("provenance loop, branch %s cannot be provenant on itself", pfsdb.BranchKey(provBranch))
 	}
 	add(&branchInfo.Provenance, provBranch)
 	provBranchInfo := &pfs.BranchInfo{}
-	if err := d.branches.ReadWrite(sqlTx).Upsert(pfsdb.BranchKey(provBranch), provBranchInfo, func() error {
-		// Set provBranch, we may be creating this branch for the first time
-		provBranchInfo.Branch = provBranch
+	if err := d.branches.ReadWrite(txnCtx.SqlTx).Upsert(pfsdb.BranchKey(provBranch), provBranchInfo, func() error {
+		if provBranchInfo.Branch == nil {
```
Comment (brycemcanally) [inline src/server/pfs/server/driver.go:1921]: Does this mean that branches can be created through the provenance parameter in create branch?

File: src/server/pfs/server/driver.go (hunk 5 above)
Comment (Tryneus) [reply src/server/pfs/server/driver.go:794]: The idea is that it doesn't contribute datums, so we would only create an output commit (and thus a job) based on input commits.  Currently in global IDs, we will create an initial empty job (unless the pipeline has input datums on creation).

File: src/server/pfs/server/driver.go (hunk 8 above)
Comment (Tryneus) [reply src/server/pfs/server/driver.go:1914]: Yes - it looks like before it would lazily create provenance on a branch that didn't yet exist, but now we eagerly instantiate it.

File: src/server/pfs/server/driver.go (hunk 5 above)
Comment (brycemcanally) [reply src/server/pfs/server/driver.go:794]: Wouldn't you need a new output commit for the reprocessing case?

File: src/server/pfs/server/driver.go (hunk 8 above)
Comment (brycemcanally) [reply src/server/pfs/server/driver.go:1914]: I guess this is fine, but it seems a bit sketchy to create the provenant branches if they don't already exist. If the prior code was not creating the branches and not erroring, then this is certainly preferable.

**Review by brycemcanally — APPROVED** [review]
lgtm
